packages/teams-js/src/internal/performanceStatistics.ts

```diff
@@ -0,0 +1,114 @@
+  sampleCount: number; // the number of frames that were processed in the duration
+  distributionBins: Uint32Array; // an array that presents counts of frames that were finished in n milliseconds: distributionBins[frameProcessingDurationInMs]=frameCount. For example, distributionBins[10] = 5 means that 5 frames were processed in 10 milliseconds.
+};
+
+export class PerformanceStatistics {
+  private currentSession: {
+    startedAtInMs: number;
+    timeoutInMs: number;
+    effectId: string;
+    frameWidth: number;
+    frameHeight: number;
+  };
+
+  private frameProcessingStartedAt: number;
+  private distributionBins: Uint32Array;
+  private sampleCount = 0;
+  private timeoutId: number;
+
+  public constructor(
+    distributionBinSize: number,
+    private report: (result: PerformanceStatisticsResult) => void, // post event to the host
+  ) {
+    if (inServerSideRenderingEnvironment()) {
+      throw errorNotSupportedOnPlatform;
+    }
+    this.distributionBins = new Uint32Array(distributionBinSize);
+  }
+
+  /**
+   * Call this function before processing every frame
+   */
+  public processStarts(effectId: string, frameWidth: number, frameHeight: number): void {
+    if (!this.suitableForThisSession(effectId, frameWidth, frameHeight)) {
+      this.reportAndResetSession(this.getStatistics(), effectId, frameWidth, frameHeight);
+    }
+    this.start();
+  }
+
+  public processEnds(): void {
+    // calculate duration of the process and record it
+    const durationInMillisecond = performance.now() - this.frameProcessingStartedAt;
+    const binIndex = Math.floor(Math.max(0, Math.min(this.distributionBins.length - 1, durationInMillisecond)));
+    this.distributionBins[binIndex] += 1;
+    this.sampleCount += 1;
+  }
+
+  private getStatistics(): PerformanceStatisticsResult {
+    if (!this.currentSession) {
+      return null;
+    }
+    return {
+      effectId: this.currentSession.effectId,
+      frameHeight: this.currentSession.frameHeight,
+      frameWidth: this.currentSession.frameWidth,
+      duration: performance.now() - this.currentSession.startedAtInMs,
+      sampleCount: this.sampleCount,
+      distributionBins: this.distributionBins.slice(),
+    };
+  }
+
+  private start(): void {
+    this.frameProcessingStartedAt = performance.now();
+  }
+
+  private suitableForThisSession(effectId: string, frameWidth: number, frameHeight: number): boolean {
+    return (
+      this.currentSession &&
+      this.currentSession.effectId === effectId &&
+      this.currentSession.frameWidth === frameWidth &&
+      this.currentSession.frameHeight === frameHeight
+    );
+  }
+
+  private reportAndResetSession(result, effectId, frameWidth, frameHeight): void {
+    result && this.report(result);
+    this.resetCurrentSession(this.getNextTimeout(effectId, this.currentSession), effectId, frameWidth, frameHeight);
+    if (this.timeoutId) {
+      window.clearTimeout(this.timeoutId);
+    }
+    this.timeoutId = window.setTimeout(
```

Comment:
references to window need to check whether or not we're in an SSR environment and now throw in those cases. We want to make sure that people calling this function from an SSR app won't cause their app to throw an error when rendering on the server.

Reply:
`inServerSideRenderingEnvironment` is called by the constructor so we can assume all the instance methods are able to use `window`

Reply:
We can avoid using setTimeout by determine the report trigger time in the processStarts if the session exceeds the threshold duration.

Reply:
updated